custom-resource-handlers/src/pgp-secret.ts

```diff
@@ -0,0 +1,135 @@
+import aws = require('aws-sdk');
+import crypto = require('crypto');
+import fs = require('fs');
+import os = require('os');
+import path = require('path');
+import util = require('util');
+
+import cfn = require('./_cloud-formation');
+import _exec = require('./_exec');
+import lambda = require('./_lambda');
+import _rmrf = require('./_rmrf');
+import { resolveCurrentVersionId } from './_secrets-manager';
+
+const secretsManager = new aws.SecretsManager();
+const ssm = new aws.SSM();
+
+export async function main(event: cfn.Event, context: lambda.Context): Promise<void> {
+  try {
+    // tslint:disable-next-line:no-console
+    console.log(`Input event: ${JSON.stringify(event)}`);
+    const attributes = await handleEvent(event, context);
+    await cfn.sendResponse(event,
+                           cfn.Status.SUCCESS,
+                           attributes.SecretArn || event.PhysicalResourceId || context.logStreamName,
+                           attributes);
+  } catch (e) {
+    // tslint:disable-next-line:no-console
+    console.error(e);
+    await cfn.sendResponse(event,
+                           cfn.Status.FAILED,
+                           event.PhysicalResourceId || context.logStreamName,
+                           { SecretArn: '' },
+                           e.message);
+  }
+}
+
+interface ResourceAttributes {
+  SecretArn: string;
+  SecretVersionId?: string;
+  ParameterName?: string;
+
+  [name: string]: string | undefined;
+}
+
+async function handleEvent(event: cfn.Event, context: lambda.Context): Promise<ResourceAttributes> {
+  const props = event.ResourceProperties;
+  let newKey = event.RequestType === cfn.RequestType.CREATE;
+
+  if (event.RequestType === 'Update') {
+    const oldProps = event.OldResourceProperties;
+    const immutableFields = ['Email', 'Expiry', 'Identity', 'KeySizeBits', 'ParameterName', 'SecretName', 'Version'];
+    for (const key of immutableFields) {
+      if (props[key] !== oldProps[key]) {
+        // tslint:disable-next-line:no-console
+        console.log(`New key required: ${key} changed from ${oldProps[key]} to ${props[key]}`);
+        newKey = true;
+      }
+    }
+  }
+
+  switch (event.RequestType) {
+  case cfn.RequestType.CREATE:
+  case cfn.RequestType.UPDATE:
+    return newKey
```

Comment:
Just to make sure I understand, if we are in `newKey` mode (and UPDATE), the physical resource ID (secret ARN) will change and then CFN will send a DELETE on the old secret will be deleted after the new one has been created. correct? Maybe worth a short comment.

Reply:
This felt obvious enough to me to not warrant a comment. But okay

Reply:
Fun stuff - there actually is an issue here in that the old & new secret both have the same name, so the CLEANUP will cause the secret to be deleted. Not nice.